src/node/lattice.rs

```diff
@@ -0,0 +1,518 @@
+use crate::node::operator_event::OperatorEvent;
+use futures::lock::Mutex;
+use petgraph::{
+    stable_graph::{EdgeIndex, NodeIndex, StableGraph},
+    visit::DfsPostOrder,
+    Direction,
+};
+use std::collections::{HashSet, VecDeque};
+use std::sync::Arc;
+
+/// `ExecutionLattice` is a data structure that maintains
+/// [`OperatorEvent`](../operator_event/struct.OperatorEvent.html)s in a directed acyclic graph
+/// according the partial order defined.
+///
+/// Events can be added to the lattice using the `add_event` function, and retrieved using the
+/// `get_event` function. The lattice requires a notification of the completion of the event using
+/// the `mark_as_compeleted` function in order to unblock dependent events, and make them runnable.
+///
+/// # Example
+/// The below example shows how to insert events into the Lattice and retrieve runnable events from
+/// the lattice.
+/// ```
+/// use erdos::node::{operator_event::OperatorEvent, lattice::ExecutionLattice};
+/// use erdos::dataflow::Timestamp;
+/// use futures::executor::block_on;
+///
+/// async fn async_main() {
+///     let mut lattice: ExecutionLattice = ExecutionLattice::new();
+///
+///     // Add two events of timestamp 1 and 2 to the lattice with empty callbacks.
+///     lattice.add_event(OperatorEvent::new(Timestamp::new(vec![1]), true, || ())).await;
+///     lattice.add_event(OperatorEvent::new(Timestamp::new(vec![2]), true, || ())).await;
+///
+///     // Retrieve the first event from the lattice.
+///     let (event_1, event_id_1) = lattice.get_event().await.unwrap();
+///
+///     // If we try to retrieve another event, we get None since we haven't marked the
+///     // completion of the event with timestamp 1.
+///     assert_eq!(lattice.get_event().await.is_none(), true);
+///
+///     // Mark the first event as completed.
+///     lattice.mark_as_completed(event_id_1).await;
+///
+///     // Now, get the second event from the lattice.
+///     let (event_2, event_id_2) = lattice.get_event().await.unwrap();
+/// }
+///
+/// fn main() {
+///     block_on(async_main());
+/// }
+/// ```
+pub struct ExecutionLattice {
```

Comment:
Since this is only used internally:
```suggestion
pub(crate) struct ExecutionLattice {
```

Reply:
Thanks, I didn't know about this. Useful!

Reply:
Hmmm, this makes the doctests in lattice.rs fail because lattice is no longer visible from erdos::.